### PR TITLE
feat: Allow sorting collections in sidebar

### LINKED ISSUE #1759: Allow sorting collections in sidebar
### Why

Now that document drag and drop is complete, as an admin I should be able to choose the order of collections in the sidebar in the same manner.

### How

Probably a new `collections.move` endpoint to match `documents.move` and a new corresponding policy as well that only allows admins in a first version.

To avoid having to pass the entire list of collections back to the server we can use the concept of fractional indexing. I'll link the library here to achieve this soon.

### PR DESCRIPTION
**Pull request by iamsaumya**
Closes #1759 

Todo
- [x]  Take care of the top cursor.
- [x]  Introduce an index attribute in the collection table.
- [x]  Create a middleware to sanitize the index` attribute of all the collections. 
- [x]  Update `collections.create` endpoint to create an index as well. 
- [x]  Create a `collections.move` endpoint 
- [x]  Add a policy to only admin to move the collection.
- [x]  Add test for `collection.move`
- [x]  Integrate with websockets.

### CODE REVIEW

**Comment by iamsaumya** [comment]
A possible algorithm to use fractional indexing, by default all the collections will have their fractional index `null`.

1. Fetch all the collections.
2. Iterate through the collections. If two or more collections have the same index, leaving a collection to reinitialize all other to `null`.
For example: if 3 collections have the same index, then reset the index of two of them to `null`.
3. Sort the collections such that all the collections with `null` index are at the bottom of the list. For example: `["23.32","32.11",null,null]`
4. Pass the collection list to a function which will create generate a new index for the collection with `null` index value.

**Comment by iamsaumya** [comment]
Tom, please let me know if the todos and algorithm look decent enough. 

If we are introducing a new attribute in the collection table then what are things I should take care of in the codebase?

**Comment by tommoor** [comment]
Looking good – note that because the "cursor" is shown below, moving to the top of the list is a special case - for documents the additional "top" cursor is here: https://github.com/outline/outline/pull/1722/files#diff-c6ce8fd8c9fd8e2796d41937514439cadff8e2d64b3306caf77a4e4fb07d9901R106

> Tom, please let me know if the todos and algorithm look decent enough.

If the existing collection indexes are null then it does introduce a lot of annoying edge-cases. For example it won't really be possible to calculate a valid index on the client without at least calculating indexes for every item above it. This makes me think that we should create indexes "just in time" when `collections.list` is called if they don't exist, so the new client code or `collections.move` never has to deal with collections lacking an index.

> If we are introducing a new attribute in the collection table then what are things I should take care of in the codebase?

- I suggest we call the column `index`, it should be text
- See docs on [migrations](https://github.com/outline/outline#migrations)
- Add the field to the [model](https://github.com/outline/outline/blob/main/server/models/Collection.js)
- Add the field to the [presenter](https://github.com/outline/outline/blob/main/server/presenters/collection.js)

**Comment by iamsaumya** [comment]
While converting the `Collections.js` to a functional component to takecare of the top cursor, I am not sure how to implement the @keydown('n') decorator. Is there a way to implement it in a function component?

**Comment by iamsaumya** [comment]
Also, I am unable to trigger the top cursor with a document. When I try to hover in-between the first document and collection name, the drop to import functionality gets triggered. 

https://user-images.githubusercontent.com/40148448/107362701-1cf5af00-6aff-11eb-93db-0310da8a083f.mp4

**Comment by tommoor** [comment]
>  I am not sure how to implement the @keydown('n') decorator. Is there a way to implement it in a function component?

I have an issue for this, but didn't expect it to be a blocker here – https://github.com/outline/outline/issues/1856 – I think the pragmatic solution is to just move that keyboard shortcut to `Layout.js` for now, it doesn't really have to live in `Collections.js`, it was just handy because the new document function already existed.


> When I try to hover in-between the first document and collection name, the drop to import functionality gets triggered.

I'm not able to reproduce this in production, any chance it's been broken on your branch?

**Comment by iamsaumya** [comment]
> any chance it's been broken on your branch?

Yeah, It was. I fixed it.

**Comment by iamsaumya** [comment]
Hey Tom, There are some warning in the console after pulling from the main branch. You might want to check on it.

**Comment by iamsaumya** [comment]
> This makes me think that we should create indexes "just in time" when collections. the list is called if they don't exist, so the new client code or collections. the move never has to deal with collections lacking an index.

Keeping your remarks in mind, I have put some thought together and here is the solution that I propose. 

All the collections irrespective of being private or public have something in common i.e `teamId`. Therefore, we can keep all the collections of a team sorted and it will take care of all the edges cases involved in a collection that going private from public or vice versa.

We can introduce an `index` column of `text` datatype with the default value `null` ( lemme know if it has to be `""`). 

In `collections.list`,

1. We will first fetch all the collections of the user's team.
1. Validate that there are no index collision or a collection with the default value.
1. If there is, then we will fix it. 
1. Usual `collections.list`.

How will the `collections.move`  work?

`collection.move` will take 2 parameters. 
1. The collection that is to be moved.
2. The collection that will be present above the new position of the dragged collection. 

`collection.move( draggedCollection , aboveCollection )`

In the backend, we will fetch all the collections of user's team and iterate through it till we find the collection just next to the `aboveCollection` in the fetched list. If  `aboveCollection`  is the last collection then the second index will be null. 

Now we have two indexes and we calculate the new fractional index and assign it to `draggedCollection`. 

If the `draggedCollection` is moved to the top in the sidebar then the `aboveCollection` will be null and now we know that `draggedCollection` has to be at the top.

**Comment by tommoor** [comment]
`collections.list` approach and migration sounds correct.

`collections.move` can be simpler, the beauty of using the fractionalIndex approach is that you don't need to send any more data than the index of the collection being moved (no need to send the "above" collection) ([reference](https://www.figma.com/blog/realtime-editing-of-ordered-sequences/)).

`collections.move` can accept a `collectionId` and `index` in the request, on the server we must check if the collection index is existing in the team context (in the case that two people move at once, or there is a private collection that the user doesn't have access to with that index), in this case we then get the above collection and generate an index between the two. These can probably be subsequent queries as this case will be very rare.

**Comment by iamsaumya** [comment]
I am writing the logic to sanitize the `index` field of the collections for `collections.list` in a middleware so that we plug it wherever we want in future.

**Comment by iamsaumya** [comment]
whoopsie! A circleci check is failing. What exactly is the failed check about and What should I do to fix it?

File: server/api/middlewares/collectionIndexing.js
Comment (tommoor) [inline server/api/middlewares/collectionIndexing.js:7]: Needs to be `ContextWithState` imported from the internal types file to fix your CI error

File: server/api/middlewares/collectionIndexing.js
Comment (tommoor) [inline server/api/middlewares/collectionIndexing.js:11]: The problem with this approach is you're pulling all the collections from the db whether they're needed or not on every request that uses the middleware. Consider we already load collections inside of `collections.list` why not perform this as a one-off when those collections are loaded just check to see if first has a `null` index property?

**Comment by iamsaumya** [comment]
Oh god please! all this time I thought the `pending` tag meant that I am waiting for your response. Had posted this review a week before and was waiting for you 🤦

File: server/api/middlewares/collectionIndexing.js
Comment (tommoor) [reply server/api/middlewares/collectionIndexing.js:11]: I think we can use the users collections from the existing query as a "hint" to whether the migration needs to happen if the first index is `null`, and then in only that case we can query _all_ collections and fix them.

**Comment by iamsaumya** [comment]
Hey Tom, Once this feat is in place, Are we going to discard listing the collections in descending order by `updatedAt` or both the options are going to coexist?

**Comment by tommoor** [comment]
The functionality will remain in the API to sort collections by various means but the UI will only allow the manual ordering

**Comment by iamsaumya** [comment]
Alright, I will update `colletions.list` to return collections in asc order by index.

**Comment by iamsaumya** [comment]
Almost done! :man_dancing:  

Have added a few more to-dos to the list, please let me know they suffice. Thanks!!

**Comment by iamsaumya** [comment]
On hold because of https://github.com/goabstract/fractional-index/issues/1

**Comment by iamsaumya** [comment]
I think we will have to sort the collections in the app using normal `sort()`. In the package, If a is `"2"` and b is `"11"` then it will throw an error ( `a>=b` error). And because we are sorting the collections using natural sort, this error is very likely to happen in the app.

**Comment by tommoor** [comment]
Yea, that sounds right – natural sort is for human-like sorting, in this case we want rigid logical machine sorting :)

**Comment by iamsaumya** [comment]
Which comment did the outline delete? :thinking:

**Comment by iamsaumya** [comment]
I modified `collections.create` endpoint to take `index` in the body. However, I did not make it a required field as it will help the API users to create a collection without bothering about `index`. 

Please guide me on how to update the API docs to show the `index` in request and responses.

**Comment by iamsaumya** [comment]
Yay :beers:, Worked on everything I think there was to work on. If I have left something, please let know :smiley:

File: app/components/Sidebar/components/CollectionLink.js
Comment (thenanyu) [inline app/components/Sidebar/components/CollectionLink.js:199]: this is a small nit — would `"auto"` be a better default value than `"all"` here

File: app/components/Sidebar/components/CollectionLink.js
Comment (thenanyu) [inline app/components/Sidebar/components/CollectionLink.js:119]: What I think you're attempting here is to temporarily hide the collection contents when a drag is active.
- modifying central state like this is a bit of a roundabout way to accomplish this goal
- instead, move the calculation for `expanded` https://github.com/outline/outline/pull/1870/files#diff-c6ce8fd8c9fd8e2796d41937514439cadff8e2d64b3306caf77a4e4fb07d9901R47 below here and simply change its definition to also include `!isCollectionDragging`

File: server/api/collections.js
Comment (tommoor) [inline server/api/collections.js:646]:
```suggestion
  await collection.update({ index });
```

File: app/scenes/CollectionNew.js
Comment (tommoor) [inline app/scenes/CollectionNew.js:49]: What do you think about moving this to the server? After all the client might not be aware of all the collections… which would result in the new collection potentially being inserted part way down the list.

I think _allowing_ `index` in the create is okay from an API pov though

File: app/stores/CollectionsStore.js
Comment (tommoor) [inline app/stores/CollectionsStore.js:54]: See comment above about moving this to the server, we'd have to query the last collection in the team

File: app/components/Sidebar/components/Collections.js
Comment (tommoor) [inline app/components/Sidebar/components/Collections.js:25]:
```suggestion

  React.useEffect(() => {
```

Nit: A little extra space between methods and `const` would be nice :)

File: server/utils/collectionIndexing.js
Comment (tommoor) [inline server/utils/collectionIndexing.js:4]:
```suggestion

export default async function collectionIndexing(teamId: string) {
```

File: server/api/collections.js
Comment (tommoor) [inline server/api/collections.js:590]: I'm not a huge fan of this much logic running in the controller method, is the repeated index check necessary?

I was previously thinking the "migration" would be a one-time thing, perhaps the repeat check should happen as part of `collections.move` before writing the new index?

**Review by tommoor — CHANGES_REQUESTED** [review]
Thanks so much for the work here 💪

File: app/components/Sidebar/components/CollectionLink.js
Comment (tommoor) [inline app/components/Sidebar/components/CollectionLink.js:116]:
```suggestion
    begin: (monitor) => {
      setTimeout(() => {
        ui.clearActiveCollection();
      }, 0);
    },
```

In my testing this needs to be moved to next frame otherwise collections below the active collection are not draggable, this seems like it _might_ be a browser quirk

File: server/api/collections.js
Comment (iamsaumya) [reply server/api/collections.js:590]: I agree, let's move this logic elsewhere. 

I think the scenario which will lead to the repeated index is when a private collection is made public. 

How about we run this logic to specifically when the visibility of collections is changed? 

I think It will be unpredictable for the user when they move a collection `X` and they see other collections move as well because of index collision.

File: app/components/Sidebar/components/CollectionLink.js
Comment (iamsaumya) [reply app/components/Sidebar/components/CollectionLink.js:199]: Done, Thanks Nan!

File: app/components/Sidebar/components/CollectionLink.js
Comment (iamsaumya) [reply app/components/Sidebar/components/CollectionLink.js:116]: Indeed, it was a bug. I should have checked it. Thanks :smile:

File: app/scenes/CollectionNew.js
Comment (iamsaumya) [reply app/scenes/CollectionNew.js:49]: Let's move this to the server. If the index is `undefined` in the request payload we will insert the collection at the bottom of the collections (this is will be the default behaviour for the app). 

If the index is specified we will also need to check if the index characters are between the ASCII characters that we allow.

I was wondering, what If the index in the req is already taken? Trello API doesn't specify its behaviour for this edge case.

Comment (tommoor) [reply app/scenes/CollectionNew.js:49]: > I was wondering, what If the index in the req is already taken?

I think this is where moving the index de-duplication out of the `collections.list` method comes in. If we have a de-dupe check as part of create and move then we can just adjust the item downward by one if that ever happens.

File: server/api/collections.js
Comment (iamsaumya) [inline server/api/collections.js:70]: I had to sort the collections manually because `order by` wasn't returning the collections in the order presented in the app.

Comment (tommoor) [reply server/api/collections.js:70]: Do you mean the postgres order wasn't working as expected? I don't think its feasible to just query all the collections and sort them in js, I can look into the postgres ordering if that's the problem.

Comment (iamsaumya) [reply server/api/collections.js:70]: Sorry to put you through this work, but I think you should take a look as well. This was my query.

```js
  const collections = await Collection.findAll({
    where: { teamId: user.teamId, deletedAt: null },
    attributes: ["id", "index", "updatedAt"],
    order: [["index", "DESC"]],
    limit: 1,
  });
```

**Comment by iamsaumya** [comment]
Moved collections don't reflect in other users sidebar, unlike other operations which do. How does this work? We should implement it for reordering collections as well.

File: server/api/collections.js
Comment (tommoor) [reply server/api/collections.js:70]: Try the following:

`order: ['index COLLATE "C"', "ASC"]],`

Using collate makes postgres sort by the byte-order ([ref](https://dba.stackexchange.com/questions/94887/what-is-the-impact-of-lc-ctype-on-a-postgresql-database)), which is what we want in this case. I'd switch it round so that new collections go to the top – they could be lost easily at the bottom, this will also match the behavior of new docs.

Comment (iamsaumya) [reply server/api/collections.js:70]: This was very informative, Thanks Tom! 

Yeah sure, the current implementation was inspired by Trello's way of creating cards. Now, Let's make the new collections go to the top.

**Comment by iamsaumya** [comment]
I grasped a bit of how `socket.io` is being used in the codebase. I Will work on it too. 

The todos for a feat req keeps increasing as I connect more dots in the codebase :muscle:. I think this is a good sign :sweat_smile:

**Comment by iamsaumya** [comment]
@tommoor please guide on what will be the best approach to let the client know that there was an index collision when `collections.create` or `collections.move` was fired.

For `collections.move` we can send the collectionId and their new index to client in the response. However, sending these details on the `collections.create` request as well feels odd. What do you think?

**Comment by iamsaumya** [comment]
Pls ignore my last message :sweat_smile: 

I think this is it. I want to roll out this feature in the March release.

File: app/components/Sidebar/components/CollectionLink.js
Comment (tommoor) [inline app/components/Sidebar/components/CollectionLink.js:81]:
```suggestion
  // Drop to reorder Collection
```

File: app/components/Sidebar/components/CollectionLink.js
Comment (tommoor) [inline app/components/Sidebar/components/CollectionLink.js:115]: I almost forgot why we did this already, needs a comment.

```suggestion
      // setTimeout is required to move this to next render frame otherwise collections
      // below the active collection are not draggable, we believe this is a browser quirk
      setTimeout(() => {
```

File: app/components/Sidebar/components/CollectionLink.js
Comment (tommoor) [inline app/components/Sidebar/components/CollectionLink.js:133]: `key` no longer used here

```suggestion
          <DropToImport collectionId={collection.id}>
```

File: app/components/Sidebar/components/CollectionLink.js
Comment (tommoor) [inline app/components/Sidebar/components/CollectionLink.js:135]: `key` not used here either

```suggestion
```

File: app/components/Sidebar/components/CollectionLink.js
Comment (tommoor) [reply app/components/Sidebar/components/CollectionLink.js:115]: Actually, not sure this is even working any more as dragging is not collapsing the active collection at all.

File: app/components/Sidebar/components/Collections.js
Comment (tommoor) [inline app/components/Sidebar/components/Collections.js:47]:
```suggestion

  const content = (
```

File: app/stores/CollectionsStore.js
Comment (tommoor) [inline app/stores/CollectionsStore.js:115]:
```suggestion
```

File: server/api/collections.js
Comment (tommoor) [inline server/api/collections.js:63]: 👍

File: server/api/collections.js
Comment (tommoor) [inline server/api/collections.js:62]:
```suggestion
      // using LC_COLLATE:"C" because we need byte order to drive the sorting
```

File: server/api/collections.js
Comment (tommoor) [inline server/api/collections.js:61]: We only need the first collection, right?

```suggestion
    limit: 1,
    order: [
```

File: server/api/collections.js
Comment (tommoor) [inline server/api/collections.js:92]: The use of this method is pretty much the only thing I'm left questioning in this PR, it feels very brute force. We already know the new `index` we're saving here, why don't we just check if it exists in the team already with a specific query rather than querying all collections in the team..

File: server/api/collections.test.js
Comment (tommoor) [inline server/api/collections.test.js:153]: We should add a test for the expectation of what happens when the index is already taken here.

File: server/api/collections.js
Comment (iamsaumya) [reply server/api/collections.js]: Actually, this is how it should have been called

```js
  const collectionsIdWithIndex = await removeIndexCollisions(
    user.teamId,
    index,
    collections
  );
```
But again, all this logic was for my previous implementation. Earlier, I thought to remove the index collision of not only the `new index` but other indexes as well. Now that I think of it I unable to create a specific scenario where this is necessary. 

I will push your suggestion in an hour if I cannot contrive an example.

File: server/api/collections.test.js
Comment (iamsaumya) [reply server/api/collections.test.js:153]: Done. Does it look good?

**Comment by tommoor** [comment]
Made some naming tweaks and flow fixes on this branch, you able to cherry-pick the commits across?

https://github.com/outline/outline/compare/iamsaumya-reorder-collection-dev?expand=1

I can't PR into your branch unfortunately.

**Review by tommoor — APPROVED** [review]
Much appreciated @iamsaumya – great work 👏

File: server/utils/removeIndexCollisions.js
Comment (tommoor) [inline server/utils/removeIndexCollisions.js:12]: This is still a bit brute-force for my liking, we really shouldn't need to query _all_ collections like this as we're always constraining collisions to a collection that's being added or updated.

Given that this is an infrequent operation I don't think it should hold up merging the PR though

File: app/components/Sidebar/components/CollectionLink.js
Comment (tommoor) [inline app/components/Sidebar/components/CollectionLink.js:21]: I think you brought across some but not all of my changes here, we should either pass `ui` as a prop or take it from `useStores`

Comment (iamsaumya) [reply app/components/Sidebar/components/CollectionLink.js:21]: Done. Fetching from `useStores`

File: server/utils/removeIndexCollisions.js
Comment (iamsaumya) [reply server/utils/removeIndexCollisions.js:12]: I am pushing an update, pls bear with me.

**Comment by iamsaumya** [comment]
Removed the brute force approach, I wrote such code thinking of too much edges cases which in reality I think will never happen. 

Thanks for the reviews and your patience 😄

File: server/api/collections.test.js
Comment (tommoor) [inline server/api/collections.test.js:202]: Is it possible to tweak the remove collision method so that the collection that gets the updated `index` is the one being acted upon, rather than the one it happens to be "next" to? 

I think that'll make a lot more sense. The `index` you provide to create/update is a clue as to where the collection should land visually rather than a need for it to be that specific character.

Comment (iamsaumya) [reply server/api/collections.test.js:202]: I made it work this way because when I send an index through API as a developer I expect the collection's index to be that specific value. I don't care about the index of other collections. So I changed the index of other collections giving precedence to the one being acted upon.

Comment (iamsaumya) [reply server/api/collections.test.js:202]: Tom, What do you think? I can always rewrite the function to behave the way you want.

Comment (tommoor) [reply server/api/collections.test.js:202]: Personally I think the specific value is not important, the thing that's important is the position in the list but I'm very conscious of the amount of iteration we've already asked you to put into the PR.

One thing I wonder that might tip your thinking is that if we went the route of modifying the incoming `index` does that simplify the event data format too also as you wouldn't need to store an array of tuples? As only one collection index would ever be changed…

Comment (iamsaumya) [reply server/api/collections.test.js:202]: Indeed, this will definitely simplify things. I will jump right into it.

Comment (iamsaumya) [reply server/api/collections.test.js:191]: Done 🎉

Comment (tommoor) [reply server/api/collections.test.js:191]: so CLEAN

Comment (iamsaumya) [reply server/api/collections.test.js:191]: Ikr! Thanks for the nudge. I don't mind iterating a hundred times if the code quality improves.

**Review by tommoor — APPROVED** [review]
We did it 👏

**Pull request merged**
